File: 通用代码/代理模式(静态)/Proxy.java
public class Proxy implements Subject{
    //要代理哪个实现类
    private Subject subject = null;
    //默认被代理者
    public Proxy(){
        this.subject = new Proxy();
    }
    /*
     *通过构造函数传递代理者
     *example:被代理者传递进来
     *public Proxy(Subject _subject){
     * this.subject = _subject;
     *}
     */
    public Proxy(Object ...object){

    }
    //实现接口中定义的方法
    public void request(){
         this.before();
         this.subject.request();
         this.after();
    }
    //预处理
    private void before(){
         //do something
    }
    //善后处理
    private void after(){
         //do something
    }
}

File: 通用代码/代理模式(C++)/Proxy.cpp
#include "Proxy.h"
#include "RealSubject.h"
#include <iostream>
using namespace std;
//默认的代理
Proxy::Proxy():_subject(NULL)
{
    cout << "Proxy..." << endl;
    this->_subject = new RealSubject();
}
//通过构造函数传递代理
Proxy::Proxy(Subject *pSubject)
{
    cout << "Proxy..." << endl;
    this->_subject = pSubject;
}
Proxy::~Proxy(void)
{
    delete this->_subject;
    this->_subject = NULL;
    cout << "~Proxy..." << endl;
}
void Proxy::Request()
{
	this->Before();
    this->_subject->Request();
	this->After();
}

void Proxy::Before()
{
    cout << "Proxy::Before" << endl;
}

void Proxy::After()
{
    cout << "Proxy::After" << endl;
}
